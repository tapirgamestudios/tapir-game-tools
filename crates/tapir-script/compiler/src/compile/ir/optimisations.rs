use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
