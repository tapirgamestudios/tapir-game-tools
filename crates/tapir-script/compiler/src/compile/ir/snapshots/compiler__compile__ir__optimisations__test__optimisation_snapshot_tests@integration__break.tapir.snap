---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/integration/break.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 2 ----
i.9 = int 0
jmp 0

---- block 0 ----
i.10 = φ(i.9 from 2, i.17 from 5)
temp.11 = i.10
storeprop temp.11, 0
temp.12 = i.10
temp.13 = int 5
temp.14 = temp.12 > temp.13
conditional_jump temp.14 3 4

---- block 3 ----
jmp 1

---- block 4 ----
jmp 5

---- block 1 ----
return 

---- block 5 ----
temp.15 = i.10
temp.16 = int 1
i.17 = temp.15 + temp.16
jmp 0


----------- optimised -------------
=======================
fn @toplevel() -> 

---- block 2 ----
i.9 = int 0
jmp 0

---- block 0 ----
i.10 = φ(i.9 from 2, i.17 from 4)
storeprop i.10, 0
temp.13 = int 5
temp.14 = i.10 > temp.13
conditional_jump temp.14 1 4

---- block 1 ----
return 

---- block 4 ----
temp.16 = int 1
i.17 = i.10 + temp.16
jmp 0
